src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs

```diff
@@ -46,30 +46,62 @@ public SagaPersistenceBehavior(ISagaPersister persister, ICancelDeferredMessages
                 return;
             }
 
-            var sagaMetadata = sagaMetadataCollection.Find(context.MessageHandler.Instance.GetType());
-            var sagaInstanceState = new ActiveSagaInstance(saga, sagaMetadata, () => DateTime.UtcNow);
+
+            var currentSagaMetadata = sagaMetadataCollection.Find(context.MessageHandler.Instance.GetType());
+
+            string targetSagaTypeString;
+            string targetSagaId;
+
+            if (context.Headers.TryGetValue(Headers.SagaType, out targetSagaTypeString) && context.Headers.TryGetValue(Headers.SagaId, out targetSagaId))
+            {
+                var targetSagaType = Type.GetType(targetSagaTypeString, false);
+
+                if (targetSagaType == null)
+                {
+                    logger.Warn($"Saga headers indicated that the message was intended for {targetSagaTypeString} but that type isn't available. Will fallback to query persister for a saga instance of type {currentSagaMetadata.SagaType.FullName} and saga id {targetSagaId} instead");
+                }
+                else
+                {
+                    SagaMetadata targetSagaMetaData;
+
+                    if (!sagaMetadataCollection.TryFind(targetSagaType, out targetSagaMetaData))
+                    {
+                        logger.Warn($"Saga headers indicated that the message was intended for {targetSagaType.FullName} but no metadata was found for that saga type. Will fallback to query persister for a saga instance of type {currentSagaMetadata.SagaType.FullName} and saga id {targetSagaId} instead");
+                    }
+                    else
+                    {
+                        if (targetSagaMetaData.SagaType != currentSagaMetadata.SagaType)
+                        {
+                            logger.Warn($"Message was intended for {targetSagaType.FullName} so {currentSagaMetadata.SagaType.FullName} will not be invoked");
```

Comment:
Good catch, it was supposed to be a DEBUG log but perhaps that's overkill as well?